pkg/controller/hibernation/azure_actuator_test.go

```diff
@@ -0,0 +1,247 @@
+			azureClient := mockazureclient.NewMockClient(ctrl)
+			setupAzureClientInstances(azureClient, test.instances)
+			if test.setupClient != nil {
+				test.setupClient(t, azureClient)
+			}
+			actuator := testAzureActuator(azureClient)
+			var err error
+			switch test.testFunc {
+			case "StopMachines":
+				err = actuator.StopMachines(testAzureClusterDeployment(), nil, log.New())
+			case "StartMachines":
+				err = actuator.StartMachines(testAzureClusterDeployment(), nil, log.New())
+			default:
+				t.Fatal("Invalid function to test")
+			}
+			assert.Nil(t, err)
+		})
+	}
+}
+
+func TestAzureMachinesStoppedAndRunning(t *testing.T) {
+	tests := []struct {
+		name        string
+		testFunc    string
+		expected    bool
+		instances   map[string]int
+		setupClient func(*testing.T, *mockazureclient.MockClient)
+	}{
+		{
+			name:      "Stopped - All machines stopped",
+			testFunc:  "MachinesStopped",
+			expected:  true,
+			instances: map[string]int{"deallocated": 3, "stopped": 2},
+		},
+		{
+			name:      "Stopped - Some machines starting",
+			testFunc:  "MachinesStopped",
+			expected:  false,
+			instances: map[string]int{"deallocated": 3, "stopped": 2, "starting": 2},
+		},
+		{
+			name:      "Stopped - machines running",
+			testFunc:  "MachinesStopped",
+			expected:  false,
+			instances: map[string]int{"running": 3, "deallocated": 2},
+		},
+		{
+			name:      "Running - All machines running",
+			testFunc:  "MachinesRunning",
+			expected:  true,
+			instances: map[string]int{"running": 3},
+		},
+		{
+			name:      "Running - Some machines starting",
+			testFunc:  "MachinesRunning",
+			expected:  false,
+			instances: map[string]int{"running": 3, "starting": 1},
+		},
+		{
+			name:      "Running - Some machines deallocated or deallocating",
+			testFunc:  "MachinesRunning",
+			expected:  false,
+			instances: map[string]int{"running": 3, "deallocated": 2, "stopped": 1, "deallocating": 3},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			azureClient := mockazureclient.NewMockClient(ctrl)
+			setupAzureClientInstances(azureClient, test.instances)
+			if test.setupClient != nil {
+				test.setupClient(t, azureClient)
+			}
+			actuator := testAzureActuator(azureClient)
+			var err error
+			var result bool
+			switch test.testFunc {
+			case "MachinesStopped":
+				result, err = actuator.MachinesStopped(testAzureClusterDeployment(), nil, log.New())
+			case "MachinesRunning":
+				result, err = actuator.MachinesRunning(testAzureClusterDeployment(), nil, log.New())
+			default:
+				t.Fatal("Invalid function to test")
+			}
+			require.Nil(t, err)
+			assert.Equal(t, test.expected, result)
+		})
+	}
+}
+
+func testAzureClusterDeployment() *hivev1.ClusterDeployment {
+	return testcd.BasicBuilder().Options(func(cd *hivev1.ClusterDeployment) {
+		cd.Spec.ClusterMetadata = &hivev1.ClusterMetadata{
+			InfraID: azureTestInfraID,
+		}
+	}).Build()
```

Comment:
```suggestion
	return testcd.Build(func(cd *hivev1.ClusterDeployment) {
		cd.Spec.ClusterMetadata = &hivev1.ClusterMetadata{
			InfraID: azureTestInfraID,
		}
	})
```
But there is not much benefit in using a builder here.
```suggestion
	return &hivev1.ClusterDeployment{
		Spec: hivev1.ClusterDeploymentSpec{
			ClusterMetdata: &hivev1.ClusterMetadata{
				InfraID: azureTestInfraID,
			},
		},
	}
```